### PR TITLE
fix(router-store): Handle navigation error, cancel/error action contain real previous state

### LINKED ISSUE #1292: Router-Store: Navigation error resulting from store initiated route change is unhandled
## Minimal reproduction of the bug/regression with instructions:

https://stackblitz.com/edit/ngrx-seed-9chqaj?file=src%2Fapp%2Fapp.component.ts

Click the button, see unhandled error in the console

## Expected behavior:

Router navigation errors resulting from a store initiated navigation should be handled. There already is a NAVIGATION_ERROR action, having it unhandled seems wrong. Instead the error could be handed to the ErrorHandler, so it just gets logged.

## Versions of NgRx, Angular, Node, affected browser(s) and operating system(s):

Latest

## I would be willing to submit a PR to fix this issue

[x] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

### PR DESCRIPTION
**Pull request by dummdidumm**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?

Fixes #1292, #1293 

- Error during navigation initiated by store is caught and handed to ErrorHandler
- `ROUTER_ERROR` / `ROUTER_CANCEL` action now contain previos router and store state 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/router-store/spec/integration.spec.ts [hunk 1]
```diff
@@ -119,7 +126,7 @@ describe('integration spec', () => {
       });
   });
 
-  it('should support rolling back if navigation gets canceled', (done: any) => {
+  it('should support rolling back if navigation gets canceled (navigation initialized through roter)', (done: any) => {
```
Comment (timdeschryver) [inline modules/router-store/spec/integration.spec.ts]: typo, roter -> router

File: modules/router-store/src/router_store_module.ts [hunk 2]
```diff
@@ -301,7 +303,8 @@ export class StoreRouterConnectingModule {
     private store: Store<any>,
     private router: Router,
     private serializer: RouterStateSerializer<SerializedRouterStateSnapshot>,
-    @Inject(ROUTER_CONFIG) private config: StoreRouterConfig
+    @Inject(ROUTER_CONFIG) private config: StoreRouterConfig,
+    private errorHandler: ErrorHandler
```
Comment (timdeschryver) [inline modules/router-store/src/router_store_module.ts]: move this to above the `@Inject` parameter.

File: modules/router-store/src/router_store_module.ts [hunk 3]
```diff
@@ -310,29 +313,31 @@ export class StoreRouterConnectingModule {
...
     }
 
-    const url = this.storeState[this.stateKey].state.url;
+    const url = routerStoreState.state.url;
     if (this.router.url !== url) {
+      this.storeState = storeState;
       this.trigger = RouterTrigger.STORE;
-      this.router.navigateByUrl(url);
+      this.router.navigateByUrl(url).catch(e => {
+        this.errorHandler.handleError(e);
```
Comment (timdeschryver) [inline modules/router-store/src/router_store_module.ts]: we should write a test for this.

File: modules/router-store/src/router_store_module.ts [hunk 4]
```diff
@@ -310,29 +313,31 @@ export class StoreRouterConnectingModule {
...
       return;
     }
 
-    const url = this.storeState[this.stateKey].state.url;
+    const url = routerStoreState.state.url;
     if (this.router.url !== url) {
+      this.storeState = storeState;
       this.trigger = RouterTrigger.STORE;
-      this.router.navigateByUrl(url);
+      this.router.navigateByUrl(url).catch(e => {
```
Comment (timdeschryver) [inline modules/router-store/src/router_store_module.ts]: rename `e` to `err`

File: modules/router-store/src/router_store_module.ts [hunk 5]
```diff
@@ -377,28 +389,31 @@ export class StoreRouterConnectingModule {
...
-        this.routerState
+        nextRouterState
       ),
     });
   }
 
   private dispatchRouterCancel(event: NavigationCancel): void {
     this.dispatchRouterAction(ROUTER_CANCEL, {
-      routerState: this.routerState,
+      routerState: this.routerState!,
```
Comment (timdeschryver) [inline modules/router-store/src/router_store_module.ts]: Is there a reason why `this.routerState!` is used?
Is it to make clear the router state can't be null at this point?

File: modules/router-store/src/router_store_module.ts (hunk 3 above)
Comment (dummdidumm) [reply modules/router-store/src/router_store_module.ts]: There already is, kind of, the "...if navigation errors (navigation initiated by store)"-test. I will enhance it to check the error handler is invoked correctly.

File: modules/router-store/src/router_store_module.ts (hunk 5 above)
Comment (dummdidumm) [reply modules/router-store/src/router_store_module.ts]: Yes, that's the only reason.

**Comment by dummdidumm** [comment]
The RouterTrigger enum indicates who initiated the navigation and who is currently in control of it. If the navigation is initiated through the router, f.e. with `router.navigate(...)`, then the trigger is `ROUTER` and changes to the router store state should not trigger another `router.navigateByUrl(..)`. If the navigation is initiated through the store, then the trigger is `STORE`, `router.navigateByUrl(...)` is invoked, but we don't want router actions changing the store state to be dispatched during that. `NONE` means that either one of them can now initiate a new cycle. That's why `this.trigger !== RouterTrigger.STORE` / `this.trigger !== RouterTrigger.ROUTER` is important.

**Pull request merged**